dynamodb/bucket.go

```diff
@@ -0,0 +1,149 @@
+package dynamodb
+
+import (
+	"sync"
+	"time"
+
+	"github.com/Clever/leakybucket"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+)
+
+var _ leakybucket.Bucket = &bucket{}
+
+type bucket struct {
+	name                string
+	capacity, remaining uint
+	reset               time.Time
+	rate                time.Duration
+	db                  *bucketDB
+	mutex               sync.Mutex
+}
+
+// Capacity ...
+func (b *bucket) Capacity() uint {
+	return b.capacity
+}
+
+// Remaining space in the bucket.
+func (b *bucket) Remaining() uint {
+	return b.remaining
+}
+
+// Reset returns when the bucket will be drained.
+func (b *bucket) Reset() time.Time {
+	return b.reset
+}
+
+// Add to the bucket.
+func (b *bucket) Add(amount uint) (leakybucket.BucketState, error) {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	dbBucket, err := b.db.bucket(b.name)
+	if err != nil {
+		return leakybucket.BucketState{}, err
+	}
+	if dbBucket.expired() {
+		dbBucket, err = b.db.flushBucket(*dbBucket, time.Now().Add(b.rate))
+		if err != nil {
+			return leakybucket.BucketState{}, err
+		}
+	}
+	// update local state
+	b.remaining = b.capacity - min(dbBucket.Value, b.capacity)
```

Comment:
The main point of the `min` is for wrap around safety since the underlying types here are all `uint`s

They'll wrap around and blow the bucket if `dbBucket.Value > b.capacity`

https://play.golang.org/p/8EK2VPcnFyk

Reply:
actually wait..my case breaks it anyways 🙈